hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcConnection.java

```diff
@@ -297,11 +295,10 @@ public void operationComplete(ChannelFuture future) throws Exception {
             established(ch);
           }
         }
-      }).channel();
+      }).sync().channel();
```

Comment:
HBase shell is not able to connect without this. I'm sure I've seen errors in RS-Master communication as well, but it works in most cases.

The issue is:
```
ERROR: Failed contacting masters after 1 attempts.
Exceptions:
java.io.IOException: Call to address=Andors-MacBook-Pro.local:16000 failed on local exception: java.io.IOException: java.lang.UnsupportedOperationException: unsupported message type: Call (expected: ByteBuf, FileRegion)
```

My thinking is that `connect()` is having a race with `sendRequest()`, becaue NettyRpcConnection is trying to establish the connection in lazy way: doesn't do anything until the first request (`Call`) comes in. Because of this we have to make sure that the Netty pipeline is in a usable state before letting sendRequest() to push the Call into and seems like `sync()` is enough. It waits for the connection to be established before we give back the channel.

This is still not ideal IMHO: we could implement a latch to wait for the SSL handshake to happen before doing anything else - like ZooKeeper does -, but it might be overkill in this case given that it already works fine with sync().

Reply:
Your explanation certainly makes sense to me. I think keeping the Channel async is probably going to give us some performance gains.

If this only affected the hbase client when TLS was on, it's ok to leave this `sync()` for now and come back to it later. However, this is going to affect all clients so we may be introducing a performance regression for this optional feature which isn't good.

You mentioned a latch for when the TLS handshake is done. I vaguely remember a similar latch for the SASL handshake (though, it might have be just a synchronous check rather than asynchronous). It's also OK for us to commit this patch onto the feature branch and address it later. We just have to remember that this would block our merge back to `master`.

Reply:
There is a BufferCallBeforeInitHandler to handle the case for writing rpc call out before initializing the connection.

https://github.com/apache/hbase/blob/master/hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/BufferCallBeforeInitHandler.java

That's why we can avoid calling sync in this method.

Reply:
@Apache9 Yep, I saw that class piles up the Calls until it receives the green signal, but for some reason SSL doesn't work if the handler precedes the SSL handler. I'll take another look.

@joshelser I'd like to commit this independently from the JWT feature if it's feasible. The PR is against the master branch intentionally.